src/parse_util.cpp

```diff
@@ -456,15 +456,65 @@ void parse_util_token_extent(const wchar_t *buff, size_t cursor_pos, const wchar
     assert(pb <= (buff + bufflen));
 }
 
-wcstring parse_util_unescape_wildcards(const wcstring &str) {
+wcstring parse_util_unescape_wildcards(const wcstring &str, bool extended) {
     wcstring result;
     result.reserve(str.size());
-    bool unesc_qmark = !feature_test(features_t::qmark_noglob);
+    bool unesc_qmark = extended || !feature_test(features_t::qmark_noglob);
 
     const wchar_t *const cs = str.c_str();
+    ssize_t open_count = 0;
     for (size_t i = 0; cs[i] != L'\0'; i++) {
-        if (cs[i] == L'*') {
+        if (extended && cs[i] == L'*' && cs[i + 1] == L'*') {
+            result.push_back(ANY_STRING_RECURSIVE);
+            i += 1;
+        } else if (cs[i] == L'*') {
             result.push_back(ANY_STRING);
+        } else if (extended && cs[i] == L'\\' && cs[i + 1] == L'[') {
+            result.push_back(cs[i + 1]);
+            i += 1;
+        } else if (extended && cs[i] == L'[') {
+            if (open_count == 0) {
+                result.push_back(WILDCARD_RANGE_START);
+                open_count++;
+            } else {
+                result.push_back(cs[i]);
+                if (open_count > 0) open_count++;
+            }
+        } else if (cs[i] == L'\\' && cs[i + 1] == L']') {
+            result.push_back(cs[i + 1]);
+            i += 1;
+        } else if (extended && cs[i] == L']') {
+            if (cs[i - 1] != L'^' && cs[i - 1] != L'!') {
+                open_count--;
+                if (open_count == 0) {
+                    result.push_back(WILDCARD_RANGE_STOP);
+                } else if (open_count < 0) {
+                    // No open bracket means it can't be a range.
+                    result.push_back(cs[i]);
+                    open_count = 0;
+                } else {
+                    // If the brackets aren't balanced,
+                    // the last ']' becomes the STOP.
+                    // So let's peek ahead.
+                    bool have_later_bracket = false;
+                    bool escaped = false;
+                    for (auto j = i + 1; j < str.size(); j++) {
+                        if (cs[j] == L'\\') escaped = !escaped;
```

Comment:
Any special glob characters, so [ ] ?, `*` and `\`

Reply:
Anyway, this bit is simply peeking ahead so we see if the `]` are balanced, so this here is *just* about escaping `]`.

The first case - `if (escaped)` seems to be superfluous - if escaped is true and it's a `\\`, then we'd be setting escaped to false *already*.

Reply:
> The first case - `if (escaped)` seems to be superfluous

Yep, I just wrote it like that because sometimes we add a `continue` to the `if (escaped)` case.